Display "Origin" column in Orders table in Orders Analytics

ecgan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes https://github.com/woocommerce/woocommerce/issues/46383.

In this PR: 

- We add a new "Origin" column into the Orders table in Orders Analytics. 
    - The `GET /wp-json/wc-analytics/reports/orders` endpoint is modified to return the origin data in `extended_info.attribution.origin` in the order objects in the API response. Example screenshot below:

<img width="367" alt="image" src="https://github.com/woocommerce/woocommerce/assets/417342/635780ac-3c96-4231-92a6-aee77dfeccf5">

- We add a new "Origin" column into the exported CSV file when users click on the "Download" button in the Orders table in Orders Analytics.
    - Note that there are two methods of generating the CSV file export: client side CSV generation when there are 25 or less records, and server side CSV generation when there are more than 25 records. See https://github.com/woocommerce/woocommerce/blob/9f9d0268cd6f601b9c3d8bdd51ddf93b11a5df41/docs/reporting/adding-columns-to-analytics-reports-and-csv-downloads.md for more info.

Screenshot below shows the new "Origin" column in the Orders table in Orders Analytics:

<img width="322" alt="image" src="https://github.com/woocommerce/woocommerce/assets/417342/4904fb67-0a2b-46b8-a462-1ccddc663fdf">

Screenshot below shows the new "Origin" column in the exported CSV file:

<img width="242" alt="image" src="https://github.com/woocommerce/woocommerce/assets/417342/9b1de587-e54a-4ebd-820a-29b47c8328b3">


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

NOTE: Orders Analytics has a cache in place. To test the code, you may refer to commit https://github.com/woocommerce/woocommerce/pull/46424/commits/1a48761a0cfd215bca68839ce8f1261fe2ad65d6 to disable caching.

1. Create some orders with different order attribution origin. You may refer to PR https://github.com/woocommerce/woocommerce/pull/44907 on how to do that. 
    - You should have less than 25 orders to test client side CSV export file generation.
    - Alternatively, you can also use the date filter in Orders Analytics page to adjust the number of orders to be displayed.
3. Go to Orders Analytics: `/wp-admin/admin.php?page=wc-admin&path=%2Fanalytics%2Forders`
4. In the Orders table at the bottom of the page, there should be a new "Origin" column. Make sure the values are correct for the orders, i.e. they should match what you see in the Orders listing page or in Order details page.
5. In the table header, click on the "Download" button. A CSV file should be generated and downloaded immediately. Check that the CSV file contains the "Origin" column, and the values are correct for the orders.
6. Create more orders with different order attribution origin. You should have more than 25 orders to test server side CSV export file generation.
7. Go to Orders Analytics: `/wp-admin/admin.php?page=wc-admin&path=%2Fanalytics%2Forders`
8. In the table header, click on the "Download" button. You should see a notification toast at the bottom that says you will receive an email with a link to download the CSV file. You may need to wait for a few seconds or minutes to get the email. In the email, click on the link to get the CSV file. Check that the CSV file contains the "Origin" column, and the values are correct for the orders.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

ecgan committed:
Display channel column in Orders Analytics UI.

ecgan committed:
Set $data to false to not use cache.

For development purpose.

ecgan committed:
Get channel info from order meta and put into extended_info.

ecgan committed:
Set channel in order item level.

ecgan committed:
Replace channel with get_origin_label in extended_info.

ecgan committed:
Remove unnneeded channel in orders_data.

ecgan committed:
Display origin instead of channel in table.

ecgan committed:
Fix lint errors.

ecgan committed:
Query order meta table based on HPOS.

ecgan committed:
Remove code for development purpose.

ecgan committed:
Add changelog.

ecgan committed:
Fix code comment.

ecgan committed:
Merge branch 'trunk' into add/orders-analytics-channel-column

ecgan committed:
Guard against null values.

`_wc_order_attribution_utm_source` can be null when the order is created via web admin.

ecgan committed:
Set default origin label to "Unknown".

This default is the same as found in `output_origin_column` function in plugins/woocommerce/src/Internal/Orders/OrderAttributionController.php.

ecgan committed:
Support server side report download.

This is needed when there are more than 25 records in the Orders Analytics report.

ecgan commented:
@layoutd , since you are the main contributor for the order attribution feature, could you help to review and test this feature please? 🙏 

Also cc my co-DRI @ibndawood , for your info.

ecgan committed:
Fix failed test.

ecgan committed:
Fix lint error.

ecgan committed:
Simplify code.

We loop through $orders_data instead of $order_attributions data. This is because $order_attributions may not have attributions data for some orders.

By looping through $orders_data, it would automatically cater for "Unknown" data from get_origin_label function call, and we don't need to specify "Unknown" as default value for origin.

ibndawood commented on plugins/woocommerce/src/Admin/API/Reports/Orders/DataStore.php:
```diff
@@ -398,11 +403,14 @@ protected function include_extended_info( &$orders_data, $query_args ) {
 				'products' => array(),
 				'coupons'  => array(),
 				'customer' => array(),
+				'origin'   => '',
```
🚧 The `attribution` property is an `object`. Hence the default should be an empty `array` instead of a `string`.
```suggestion
				'products'    => array(),
				'coupons'     => array(),
				'customer'    => array(),
				'attribution' => array(),
```

ibndawood commented on plugins/woocommerce/src/Admin/API/Reports/Orders/Controller.php:
```diff
@@ -229,6 +229,12 @@ public function get_item_schema() {
 						'context'     => array( 'view', 'edit' ),
 						'description' => __( 'Order customer information.', 'woocommerce' ),
 					),
+					'channel'  => array(
+						'type'        => 'object',
+						'readonly'    => true,
+						'context'     => array( 'view', 'edit' ),
+						'description' => __( 'Order channel information.', 'woocommerce' ),
+					),
```
🚧 Per the screenshot below, we'd need more than one piece of information. Hence, we'd need to keep this property name as `attribution,` which would be an `object` with properties like `origin`, `device`, etc.

![image](https://github.com/woocommerce/woocommerce/assets/179533/2ddfce5c-fe46-4347-941a-100760fe7e74)

```suggestion
					'attribution'  => array(
						'type'        => 'object',
						'readonly'    => true,
						'context'     => array( 'view', 'edit' ),
						'description' => __( 'Order attribution information.', 'woocommerce' ),
					),
```

ecgan replied:
Oh right, I missed that screenshot. I thought it was just one "Origin" column, I didn't notice the "Device type" column. Thanks for checking that.

ibndawood commented on plugins/woocommerce/src/Admin/API/Reports/Orders/DataStore.php:
```diff
@@ -398,11 +403,14 @@ protected function include_extended_info( &$orders_data, $query_args ) {
 				$orders_data[ $key ]['extended_info']['customer'] = $mapped_customers[ $order_data['customer_id'] ];
 			}
+
+			$orders_data[ $key ]['extended_info']['origin'] = $this->get_origin_label( $order_attributions[ $order_data['order_id'] ]['_wc_order_attribution_source_type'] ?? '', $order_attributions[ $order_data['order_id'] ]['_wc_order_attribution_utm_source'] ?? '' );
```
🚧 Should change the index here as well.

ecgan replied:
Done in https://github.com/woocommerce/woocommerce/pull/46424/commits/e4963871b8772ca8a8c372830d592555cff1cf96.

ibndawood commented on plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/reports/class-wc-tests-reports-orders.php:
```diff
@@ -110,6 +110,7 @@ public function test_extended_info() {
 						),
 						'coupons'  => array(),
 						'customer' => $data->data[0]['extended_info']['customer'], // Not under test.
+						'origin'   => 'Unknown',
```
🚧 Should change the index here as well.

ecgan replied:
Done in https://github.com/woocommerce/woocommerce/pull/46424/commits/e1589b7e36bce1b817e7e655a9e6c9c2003afe72.

ibndawood commented on plugins/woocommerce/src/Admin/API/Reports/Orders/DataStore.php:
```diff
@@ -534,6 +542,52 @@ protected function get_coupons_by_order_ids( $order_ids ) {
+
+		return $order_attributions;
+	}
+
```
📜 

I took quite a bit of time to analyse this approach. I was inclined to use `wc_get_orders` instead of direct SQL. The advantages of using `wc_get_orders` are that we don't have to worry about `HPOS` compatibility, we have the `WC_Order` object giving us more information to work with, and our code is a bit cleaner. 

I also reached out to @vedanshujain and asked his opinion. He said both approaches are acceptable.

I did a simple benchmark test using `microtime`, and I can see using direct SQL is slightly faster than `wc_get_orders`. In the context of analytics, I am inclined towards faster results than cleaner code. Let's move forward with this approach and we can tweak it if required at a latest stage.

ibndawood reviewed (changes requested):
Hey @ecgan, I've added my review with some feedback. Please take a look and let me know what do you think.

ibndawood commented on plugins/woocommerce/src/Admin/API/Reports/Orders/DataStore.php:
```diff
@@ -534,6 +542,52 @@ protected function get_coupons_by_order_ids( $order_ids ) {
+	 */
+	protected function get_order_attributions_by_order_ids( $order_ids ) {
+		global $wpdb;
+		$order_meta_table   = OrdersTableDataStore::get_meta_table_name();
+		$included_order_ids = implode( ',', $order_ids );
+
+		if ( OrderUtil::custom_orders_table_usage_is_enabled() ) {
+			/* phpcs:disable WordPress.DB.PreparedSQL.InterpolatedNotPrepared */
+			$order_attributions_meta = $wpdb->get_results(
+				"SELECT order_id, meta_key, meta_value
+					FROM $order_meta_table
+					WHERE order_id IN ({$included_order_ids})
+					AND meta_key IN ('_wc_order_attribution_source_type', '_wc_order_attribution_utm_source')
+					",
+				ARRAY_A
+			);
+			/* phpcs:enable */
+		} else {
+			/* phpcs:disable WordPress.DB.PreparedSQL.InterpolatedNotPrepared */
+			$order_attributions_meta = $wpdb->get_results(
+				"SELECT post_id as order_id, meta_key, meta_value
+					FROM {$wpdb->prefix}postmeta
+					WHERE post_id IN ({$included_order_ids})
+					AND meta_key IN ('_wc_order_attribution_source_type', '_wc_order_attribution_utm_source')
+					",
+				ARRAY_A
+			);
+			/* phpcs:enable */
+		}
+
+		$order_attributions = array();
+		foreach ( $order_attributions_meta as $meta ) {
+			if ( ! isset( $order_attributions[ $meta['order_id'] ] ) ) {
+				$order_attributions[ $meta['order_id'] ] = array();
+			}
+			$order_attributions[ $meta['order_id'] ][ $meta['meta_key'] ] = $meta['meta_value'];
+		}
+
+		return $order_attributions;
+	}
```
📜 

I took quite a bit of time to analyse this approach. I was inclined to use `wc_get_orders` instead of direct SQL. The advantages of using `wc_get_orders` are that we don't have to worry about `HPOS` compatibility, we have the `WC_Order` object giving us more information to work with, and our code is a bit cleaner. 

I also reached out to @vedanshujain and asked his opinion. He said both approaches are acceptable.

I did a simple benchmark test using `microtime`, and I can see using direct SQL is slightly faster than `wc_get_orders`. In the context of analytics, I am inclined towards faster results than cleaner code. Let's move forward with this approach and we can tweak it if required at a later stage. 

What are your thoughts? 

cc: @vedanshujain

ecgan replied:
> I was inclined to use `wc_get_orders` instead of direct SQL. [...]

Yes, I thought about using `wc_get_orders` too (reference: https://github.com/woocommerce/woocommerce/wiki/wc_get_orders-and-WC_Order_Query), and I was also thinking about HPOS compatibility.

However, there were two things that I thought about, that I ended up using direct SQL.

One, I'm not sure if `wc_get_orders` might give us too much data ("fat" orders) and cause us issues in production. In our development environment, it's usually pretty simple small order objects; but I'm concerned that in the real world, merchants might be using some plugins or extensions that stores a lot of metadata into an order object. If we use `wc_get_orders`, it would be the first time we retrieve orders object in this Orders Analytics page. Even if the users choose page size 25 (the smallest page size), I'm concerned that it might cause performance issue or "out of memory" issue etc. In other words, I would like to be safe, and use direct SQL to retrieve just the data we need quickly.

Two, I see that we already have existing prior art of using direct SQL in populating other `extended_info` properties (product, customer, and coupon), so I just follow the existing way. 

> I did a simple benchmark test using `microtime` [...]

Just sharing an info that I found in https://www.php.net/manual/en/function.microtime.php: "For performance measurements, using [hrtime()](https://www.php.net/manual/en/function.hrtime.php) is recommended."

But no worry, I guess using `microtime` is fine as a rough gauge. 

> [...] and I can see using direct SQL is slightly faster than `wc_get_orders`.

Related to my comments above about "fat" orders. Using direct SQL is slightly faster in our _development environment_. I'm thinking that in the real world merchant's WooCommerce store production environment, it may be significantly faster compared to using `wc_get_orders`.

ecgan committed:
Change "channel" to "attribution" object in controller.

This is because we need more than one piece of information, so we need it to be an object with properties like `origin`, `device` etc.

Co-authored-by: Kader Ibrahim S <kader.ibrahim.s@a8c.com>

ecgan committed:
Change `origin` string to `attribution` array in Orders DataStore.

Co-authored-by: Kader Ibrahim S <kader.ibrahim.s@a8c.com>

ecgan committed:
Change origin string to attribution object.

ecgan committed:
Fix indexing after changing from origin string to attribution object.

ecgan committed:
Change from origin string to attribution object in table.js.

ecgan committed:
Simplify code.

To make code shorter and easier to read, instead of long horizontal code.

ecgan committed:
Fix lint errors.

ecgan committed:
Fix failed test.

ecgan committed:
Fix lint error.

ecgan commented:
@ibndawood , thanks for the review. I have addressed your comments and make the necessary changes. Could you do another round of review please?

ecgan committed:
Fix retrieving origin in CSV export.

ibndawood commented on plugins/woocommerce/src/Admin/API/Reports/Orders/DataStore.php:
```diff
@@ -534,6 +546,52 @@ protected function get_coupons_by_order_ids( $order_ids ) {
+			/* phpcs:enable */
+		} else {
+			/* phpcs:disable WordPress.DB.PreparedSQL.InterpolatedNotPrepared */
+			$order_attributions_meta = $wpdb->get_results(
+				"SELECT post_id as order_id, meta_key, meta_value
+					FROM {$wpdb->prefix}postmeta
+					WHERE post_id IN ({$included_order_ids})
+					AND meta_key IN ('_wc_order_attribution_source_type', '_wc_order_attribution_utm_source')
+					",
+				ARRAY_A
+			);
```
💅 A cosmetic change
```suggestion
			$order_attributions_meta = $wpdb->get_results(
				"SELECT post_id as order_id, meta_key, meta_value
					FROM $wpdb->postmeta
					WHERE post_id IN ({$included_order_ids})
					AND meta_key IN ( '_wc_order_attribution_source_type', '_wc_order_attribution_utm_source' )
					",
				ARRAY_A
			);
```

ecgan replied:
Thanks! I forgot about the `$wpdb->postmeta`, and I actually copied the `{$wpdb->prefix}postmeta` from another file. 😄

ibndawood commented on plugins/woocommerce/src/Admin/API/Reports/Orders/DataStore.php:
```diff
@@ -534,6 +546,52 @@ protected function get_coupons_by_order_ids( $order_ids ) {
+
+		if ( OrderUtil::custom_orders_table_usage_is_enabled() ) {
+			/* phpcs:disable WordPress.DB.PreparedSQL.InterpolatedNotPrepared */
+			$order_attributions_meta = $wpdb->get_results(
+				"SELECT order_id, meta_key, meta_value
+					FROM $order_meta_table
+					WHERE order_id IN ({$included_order_ids})
+					AND meta_key IN ('_wc_order_attribution_source_type', '_wc_order_attribution_utm_source')
+					",
+				ARRAY_A
+			);
```
💅 A cosmetic change
```suggestion
			$order_attributions_meta = $wpdb->get_results(
				"SELECT order_id, meta_key, meta_value
					FROM $order_meta_table
					WHERE order_id IN ({$included_order_ids})
					AND meta_key IN ( '_wc_order_attribution_source_type', '_wc_order_attribution_utm_source' )
					",
				ARRAY_A
			);
```

ibndawood commented on plugins/woocommerce/src/Admin/API/Reports/Orders/DataStore.php:
```diff
@@ -534,6 +546,52 @@ protected function get_coupons_by_order_ids( $order_ids ) {
+	protected function get_order_attributions_by_order_ids( $order_ids ) {
+		global $wpdb;
+		$order_meta_table   = OrdersTableDataStore::get_meta_table_name();
+		$included_order_ids = implode( ',', $order_ids );
```
🚧 We should sanitize the IDs
```suggestion
		$included_order_ids = implode( ',', array_map( 'absint', $order_ids ) );
```

ecgan replied:
Thanks! I actually got the `implode( ',', $order_ids );` from the other functions (e.g. `get_products_by_order_ids`). 

Anyway, I have applied your suggestion. Good to be defensive here.

ibndawood reviewed (approved):
Thank you, @ecgan, for the changes. I've tested the PR with and without HPOS. I've also followed the test instructions. The changes test well. I've added one input and some cosmetic changes. I've approved the PR. 🚀

ecgan committed:
Use $wpdb->postmeta; cosmetic change.

Co-authored-by: Kader Ibrahim S <kader.ibrahim.s@a8c.com>

ecgan committed:
Cosmetic change.

Co-authored-by: Kader Ibrahim S <kader.ibrahim.s@a8c.com>

ecgan committed:
Sanitize order IDs by using absint.

Co-authored-by: Kader Ibrahim S <kader.ibrahim.s@a8c.com>

ecgan committed:
Merge branch 'trunk' into add/orders-analytics-channel-column

ecgan committed:
Merge branch 'trunk' into add/orders-analytics-channel-column

ecgan commented:
Note that the "Origin" column is changed to "Attribution" column in PR https://github.com/woocommerce/woocommerce/pull/46760.
